Source/Dafny/DafnyAst.cs

```diff
@@ -6709,6 +6709,34 @@ public UpdateStmt(IToken tok, IToken endTok, List<Expression> lhss, List<Assignm
     }
   }
 
+  public class AssignOrReturnStmt : ConcreteUpdateStatement
+  {
+    public readonly Expression Rhs; // this is the unresolved RHS, and thus can also be a method call
+    public Statement ResolvedStmt = null;  // filled in during resolution
+
+    public override IEnumerable<Statement> SubStatements {
+      get { yield return ResolvedStmt; } // TODO does this need a null-check?
```

Comment:
I don't remember. But you may want to change `ResolveStmt` from being a single statement to a list of statements. This will give more flexibility (which I don't know if it's needed yet or not). If so, you can do what is done in `UpdateStmt`:

    public readonly List<Statement> ResolvedStatements = new List<Statement>();  // contents filled in during resolution
    public override IEnumerable<Statement> SubStatements {
      get { return ResolvedStatements; }
    }

(Still, I don't remember if `SubStatements` is ever called before successful resolution.)

Reply:
At the moment I'm using a `BlockStmt` to pack several statements into this one `ResolvedStmt`, but using a list might indeed be nicer, because then I can use `List.Add` to build it, and then when I want to implement `:-` without LHS, the `Extract()` call will not be needed, so I just do one fewer `List.Add`, nice.